src/spi.rs

```diff
@@ -1335,3 +1542,17 @@ impl_spi!(SPI3: spi_host_device_t_SPI3_HOST);
 impl_spi_any_pins!(SPI2);
 #[cfg(any(esp32, esp32s2, esp32s3))]
 impl_spi_any_pins!(SPI3);
+
+// See https://github.com/espressif/esp-idf/blob/d2471b11e78fb0af612dfa045255ac7fe497bea8/components/driver/include/esp_private/spi_common_internal.h#L464
+const MAX_SPI_DEVICES: usize = 6;
+
+#[cfg(not(any(esp32, esp32s2, esp32s3)))]
+static NOTIFIER: [MultiNotification<{ MAX_SPI_DEVICES }>; 2] =
+    [MultiNotification::new(), MultiNotification::new()];
+
+#[cfg(any(esp32, esp32s2, esp32s3))]
+static NOTIFIER: [MultiNotification<{ MAX_SPI_DEVICES }>; 3] = [
+    MultiNotification::new(),
+    MultiNotification::new(),
+    MultiNotification::new(),
+];
```

Comment:
No way I think. The ESP IDF SPI driver callbacks only take a `*mut spi_transaction_t` parameter. No user context, no nothing. What saves my a** is only that I know - from the `spi_transaction_t` instance - for which SPI unit/hardware we are talking about, and hence - I can awake all devices which are waiting on that unit/hardware/bus. They share a single Notification of course. 

Also... even if these are statics... It is not like making these non-static is going to give the user some sort of extra freedom or anything, isn't it. You have a hard (as in hardware) limit of 3 (or 2 - for some MCUs) instances of `SpiDriver` anyway.

Reply:
> Also... even if these are statics...

Yes, you are right. I just have a thing against global state in general. I've been bitten quite a few times by them in the past.

> The ESP IDF SPI driver callbacks only take a *mut spi_transaction_t parameter. No user context, no nothing.

Well I've got some good news! There is a user context. I even make use of it in my PR. It's the `spi_transaction_t.user` field. You can set it before you send the transaction and read it in the callback.

I hope this helps you remove the statics. I'd be ecstatic if it does.

Reply:
Don't get too ecstatic yet. :-) Here's another hurdle: either the driver should be `Pin`-ed right after its creation (and I hope you realize we don't want to explain Pin-ing to users), or we need `alloc` to un-"staticfy" the Notifications (as I need to Box` them, obviously). And `alloc` just for this is kind of even dirtier IMO.

What is this thing between you and statics? These are not like "Java" statics, you know. :-) And these are also private state...

Reply:
(Not that ESP IDF doesn't allocate like mad behind the curtains, but yeah, we can do better than ESP IDF, even if we are just wrappers on top of it.)

Reply:
Lmao at Java. My pain is from C++. [This](https://en.cppreference.com/w/cpp/language/siof#:~:text=C%2B%2B%20language%20The%20static%20initialization%20order%20fiasco%20refers,duration%20in%20different%20translation%20units%20are%20initialized%20in.) is one of them. I always prefer to have user control the lifetime of things.

> we need alloc to un-"staticfy" the Notifications

Can't it go on the stack (Or at least the fake stack Rust makes for async functions)? Rust will pin it for us.

Reply:
> Can't it go on the stack (Or at least the fake stack Rust makes for async functions)? Rust will pin it for us.

Well, I think calling the nested enum-based state machines to which the `async/await` syntax is lowered by the Rust generator framework a "stack" is probably pushing it too far :-), but seriously - can't think of a simple way and not sure if it is possible at all. For the notification to end up on that "stack" (or as as variable in a generated enum, that survives enum state machine transitions) it needs to be an owned member of the future returned by `read_async` / `write_async` and furthermore, the whole future should be !Unpin I think. One problem here is that notifications have a lifetime that currently co-incides with the lifetime of the driver itself, not with the lifetime of the future. Making these having a shorter lifetime means constant subscriptions/unsubscriptions from/to the notifications on future creation/drop - just to have the Notification non-static - which while giving a small fuzzy warm feeling I still think is just not worth the complexity.

> Lmao at Java. My pain is from C++. [This](https://en.cppreference.com/w/cpp/language/siof#:~:text=C%2B%2B%20language%20The%20static%20initialization%20order%20fiasco%20refers,duration%20in%20different%20translation%20units%20are%20initialized%20in.) is one of them. I always prefer to have user control the lifetime of things.

This is not a problem in Rust; at least not for stuff that does **not** rely on `lazy_static` and friends. Non-`mut` globals in Rust which are initialized with `const new() -> Self` constructors are actually nice, good citizens. Their initialization is happening before the program even starts, by the image loader copying the contents of the "elf" rwdata section into memory with a simple, single `memcpy`. Our `Notification` is one of these.

Reply:
To add on: I think - besides looking into how `static foo: X = X::new()` actually works, where `X::new()` is obviously a `const fn` or else the thing won't compile - you should also look at `static_cell` as an alternative to the `lazy_static` madness. `static_cell` gives you a *predictable*, safe, one-time initialization at runtime *that you need to explicitly call* for stuff which cannot be expressed completely with `const fn`s, yet you still get the `'static` lifetime, which is so beneficial in various cases - amongst which - essentially pinning the initialized structure post-initialization (as it is non-`mut` and hence cannot be swapped out).

Who uses `lazy_static!` on embedded anyway?

Reply:
> can't think of a simple way and not sure if it is possible at all

I'll propose a solution to this when I get back from work.

> Making these having a shorter lifetime means constant subscriptions/unsubscriptions from/to the notifications on future creation/drop

What do you mean by "constant subscriptions/unsubscriptions"? Aren't these really cheap or is there some expense that I'm missing?

Reply:
Code complexity. And that's just what I see as an issue before even starting to implement this.

Reply:
But then again: why are you so fixated on these statics? I think I just debunked your myth that their initialization is unpredictable. So then what's left?

Reply:
Don't worry about the complexity, I'll propose the solution I used in my PR and you can judge.

What myth are you talking about? The initialisation fiasco one? That's a C++ problem.

I stated at the start that I don't like global state managed by libraries and that's my opinion. Regardless of whether the state is private or public, regardless of how predictable the compiler is, regardless of how much the user cares.

If you thoroughly enjoy using global state instead of creating structs on the stack where they are needed, then you can do that if you want.

If you like things to be on the stack (like me) but find it difficult or can't be bothered to do it here, then allow me to propose some solution (or also give up in the process). We don't need to deconstruct or "debunk" why I don't like statics if you feel the same way as well.

Reply:
Fair enough! Waiting for you proposal.

Reply:
```Rust
let notification = Notification::new();
let mut transaction = Self::create_transaction(
    handle,
    read,
    write,
    transaction_length,
    rx_length,
    keep_cs_active,
)?;

transaction.user = core::ptr::addr_of!(notification) as _;

spi_device_queue_trans(handle, &mut transaction as *mut _, delay::NON_BLOCK);

notification.wait().await;

spi_device_get_trans_result(...);
```

```Rust
unsafe extern "C" fn post_transaction(transaction: *mut spi_transaction_t) {
    let transaction = unsafe { transaction.as_mut() }.unwrap();

    let notification_ptr = transaction.user as *const Notification; // TODO: Null check for sync
    let notification = unsafe { notification_ptr.as_ref() }.unwrap();

    notification.notify();
}
```

Reply:
Might actually work as long as the above future is `!Unpin`. Need to check that...

Reply:
[The book](https://rust-lang.github.io/async-book/04_pinning/01_chapter.html) claims so, so you might be onto something!

Reply:
Done now. Need to tighten with `addr_of!` but the rest is as You Had Foreseen.